Add specialized epix_slide for epi_slide_opt

brookslogan opened the pull request:
### Checklist

Please:

- [x] Make sure this PR is against "dev", not "main" (unless this is a release
      PR).
- [x] Request a review from one of the current main reviewers:
      brookslogan, nmdefries.
- [ ] Makes sure to bump the version number in `DESCRIPTION`. Always increment
      the patch version number (the third number), unless you are making a
      release PR from dev to main, in which case increment the minor version
      number (the second number).
- [ ] Describe changes made in NEWS.md, making sure breaking changes
      (backwards-incompatible changes to the documented interface) are noted.
      Collect the changes under the next release number (e.g. if you are on
      1.7.2, then write your changes under the 1.8 heading).
- See [DEVELOPMENT.md](DEVELOPMENT.md) for more information on the development
  process.

### Change explanations for reviewer

### Magic GitHub syntax to mark associated Issue(s) as resolved when this is merged into the default branch

- Resolves #609.

### TODO

- [ ] Address review comments
- [ ] Address change in `epi_slide_opt(.before = Inf)` behavior from refactor (CHECK check not running at the moment despite some others running?)

brookslogan committed:
refactor: hoist some epi_slide_opt pre-processing to helpers

For re-use in an epix_epi_slide_opt

brookslogan committed:
feat: WIP epix_epi_slide_opt

brookslogan commented:
- [x] handle all possible .f, before
- [x] finish docs, tests
- [-] port map_ea at least, maybe base on iterators
  - handle in #173 or beyond
- [x] in-source todos
- [ ] check whether frollapply on function that errors on ref time values not in `ref_time_values` will error in current scheme; consider adding skipping layer for it or documenting

brookslogan committed:
epix_epi_slide_opt: don't re-order input columns in output

brookslogan committed:
feat: support before = Inf in epix_epi_slide_opt

brookslogan committed:
feat(epi_slide_opt): improve feedback when .f is forgotten

brookslogan committed:
fix(epix_epi_slide_opt): support ... forwarding

brookslogan committed:
Clean up some comments&code + note bug

brookslogan committed:
fix(epix_epi_slide_opt): on data.table `.f`s with `.align != "right"`

brookslogan committed:
Clear out some more comments

Changing min() to [[1L]] or min & max to range don't seem to make a notable
performance difference, and require extra assumptions / uglier code.

brookslogan committed:
Fix missing ukey_names arg

brookslogan committed:
Check for missing & improper ukey_names args

brookslogan committed:
Remove some commented ideas that aren't quick wins

brookslogan committed:
WIP cleaning up approx_equal

brookslogan committed:
fix(apply_compactify): avoid arrange on data.table, `i` parsing issues

brookslogan committed:
perf: speed up compactification with `approx_equal`

brookslogan committed:
fix(approx_equal): missing import

brookslogan committed:
docs(approx_equal): roxygen2 + comment on inconsistencies/bugs

brookslogan committed:
fix(approx_equal): consistency with vec_slice(na_equal=FALSE)

brookslogan committed:
fix(approx_equal): on bare numeric matrices

brookslogan committed:
feat: approx_equal on lists

brookslogan committed:
docs(approx_equal): iterate on @return + doc approx_equal0

brookslogan committed:
WIP docs(epix_epi_slide_opt_one_epikey): initial

brookslogan committed:
Clean up unnecessary comments and unused helper functions, +@keywords internal

brookslogan committed:
approx_equal: make "abs_tol=" mandatory, +validation, +docs

brookslogan committed:
Expand epi_slide_opt_archive_one_epikey example

brookslogan committed:
WIP epi_slide_opt.epi_archive tests

brookslogan committed:
refactor: move epi_slide_opt & helpers to its own file

brookslogan committed:
Actually turn epi_slide_opt into S3 method

brookslogan committed:
More WIP on tests

brookslogan committed:
Mark renaming TODO on approx_equal

brookslogan committed:
fix!: as_epi_archive(tibble) key setting; + distrust key if data.table

brookslogan committed:
Make epi_archive key order geo !!!other time version

brookslogan committed:
docs(new_epi_archive): roxygen2 for new param requirements

brookslogan committed:
fix(epi_slide_opt.epi_archive): as.data.table(tibble) key setting

brookslogan committed:
tests(epi_slide_opt): on example data sets

brookslogan committed:
Fix & test epi_slide_opt.grouped_epi_archive behavior

brookslogan committed:
Rename approx_equal -> vec_approx_equal

brookslogan committed:
Fix missing n_groups import + epiprocess::: CHECK lint

brookslogan committed:
Address additional lints, CHECK issues

brookslogan committed:
Address missing ::: in tests

brookslogan committed:
Fix another missing `:::`

brookslogan committed:
Fix + add NEWS.md entries

We actually coincidentally caught the tibble-to-DT key-setting issue with code
intended for addressing a separate issue with `data.table`s passed in.

brookslogan committed:
Fix CHECK doc line length lint

brookslogan committed:
Fix missing library(dplyr) in example

brookslogan committed:
Fix {epiprocess} -> `{epiprocess}` in roxygen

brookslogan committed:
docs: add vec_approx_equal to pkgdown reference index

nmdefries reviewed (commented):
Various nits and some questions. The logic looks fine to me, but given the complexity of this change, someone else should review as well.

brookslogan committed:
docs(vec_approx_equal): mention vctrs::vec_proxy_equal

brookslogan commented on R/patch.R:
```diff
@@ -0,0 +1,348 @@
+  # We have five types of rows in combined_tbl:
+  # 1. From earlier_snapshot, no matching ukey in later_tbl (deletion; turn vals to
+  #    NAs to match epi_archive format)
+  # 2. From earlier_snapshot, with matching ukey in later_tbl (context; exclude from
```
It gives context for what to do with the corresponding row in `later_tbl`, which will be category 3/4.  `combined_tbl` has earlier followed by later, so ` vec_duplicate_id(combined_ukeys)` will point from the category 3/4 row back to the corresponding category 2 row; we can then look at the measurement cols to decide whether we're in case 3. or in case 4.  The category 2. row should be thrown out in either case; if the later row's category 3 then we should not have this ukey in the diff result at all, and if the later row's category 4, then we should only have that later row in the diff result.

brookslogan commented on R/patch.R:
```diff
@@ -0,0 +1,348 @@
+    ))
+  }
+
+  combined_tbl <- vec_rbind(earlier_snapshot, later_tbl)
```
Performance was one, with a `full_join`-based approach taking about 4x the time of this approach.  (That `full_join` approach didn't consider compactification tol, which means considering it would probably make it a bit slower still.)

I had a pretty fast, pretty legible approach in terms of anti joins that used a simpler form of this duplicate-logic to make a faster anti join.  Unfortunately, I had troubles implementing tolerance in an efficient way there, but could take another look if this is too much gibberish.

dsweber2 replied:
If we're keeping this in this format, it might be a good idea to directly test it, e.g. against an implementation that just uses `full_join`s on some randomly generated `earlier_snapshot` `later_tbl` pairs.

Alternatively, since we'll be compactifying later, the anti-join version that only tests direct equality might work fine. Occasionally including redundant values is less of a problem than excluding unique ones.

I do think you did a pretty good job of documenting it!

brookslogan replied:
Good idea on the testing.

Regarding compactifying later... that might be a valid approach, though this actually this doesn't compactify later, since it should be redundant.  I was actually thinking that eventually we could try compactifying by splitting a DT apart by versions, then diffing consecutive versions & recombining.  (Looks more elegant with helper functions.)   That would give us more flexibility to do things like https://github.com/cmu-delphi/epiprocess/issues/642.

brookslogan replied:
I've rewritten this in terms of anti joins & put the complex vctrs duplicate approaches inside a fast anti-join-with-tolerance function.  I think it reads better, and this iteration of it seems to actually be faster than the 5-case approach (while an older version of it was slower).  Does this seem less convoluted?

dsweber2 replied:
I definitely think so!

brookslogan commented on R/epi_slide_opt_archive.R:
```diff
@@ -0,0 +1,221 @@
+      slide_t_max <- max(inp_snapshot$time_value)
+    } else {
+      # If the input had updates in the range t1..t2, this could produce changes
+      # in slide outputs in the range t1-after..t2+before, and to compute those
+      # slide values, we need to look at the input snapshot from
+      # t1-after-before..t2+before+after. nolint: commented_code_linter
```
Suppose `before = 6, after = 0` and there's an update to the 1d value at time t alone.  We need to compute the trailing 7dav for t..t+6.  But in order to compute the trailing 7dav for time t, we need 1d data from t-6..t.  So we end up needing the union of the ranges t-6..t, t-5..t+1, t-4..t+2, ..., t..t+6, which gives us t-6..t+6.

brookslogan commented on R/epi_slide_opt_archive.R:
```diff
@@ -0,0 +1,221 @@
+          if (after != 0L) {
+            # data.table always puts NAs at tails, even with na.rm = TRUE; chop
+            # off extra NAs from beginning and place missing NAs at end:
+            out_col <- c(out_col[seq(after + 1L, slide_nrow)], rep(NA, after))
```
Nice catch!  I think I was worried about some garbage-rather-than-NA being in the tails which is why I didn't just copy that approach, but on second look that worry seems unfounded.

brookslogan replied:
Hmm, actually moving the first `after` values to the end of the result seems like it might not impact the output at all, at least for `epi_slide_opt.epi_df`, as these appear to all be non-`.real` (corresponding to padding `time_value`s) and excluded from the final result anyway.  We may need to intercept & manually handle `fill =`, or disallow it.

nmdefries replied:
I do think `fill` is convenient, but it's both not hard to implement (on our side or by the user) and doesn't change the results very much.

nmdefries replied:
See https://github.com/cmu-delphi/epiprocess/pull/611#discussion_r2015320321

brookslogan commented on R/epi_slide_opt_archive.R:
```diff
@@ -0,0 +1,221 @@
+    prev_inp_snapshot <<- inp_snapshot
+    prev_out_snapshot <<- out_snapshot
+    out_diff$version <- version
+    out_diff
```
Perhaps; some thoughts:
- Recording all full `out_snapshot`s seems like it will potentially give a very hefty result, and they could get that same result by something like `epix_slide(~ .x)` on the result or a helper function doing something like that.
- Recording just `out_snapshot` restricted to times that were in `inp_update` plus any times that are required for validity (e.g., an change in just the time t 1d value producing an update to the trailing 7dav for t+1..t+6 as well as t) seems maybe more useful.  I'm not sure how hard/slow this would be; I had some issues trying to selectively apply `compactify_tol` to the 7dav/etc. cols and set aside that idea.

brookslogan replied:
Just pushing something to disallow `fill =` altogether.  Refiling to #646.

brookslogan committed:
Fix & test some vec_approx_equal vs. vec_equal behaviors

brookslogan committed:
fix(vec_approx_equal): don't assume inds1 and inds2 non-NULL together

brookslogan committed:
perf(vec_approx_equal): avoid vec_cast_common, ptype2 work

brookslogan committed:
docs(vec_approx_equal): convert comment to issue

brookslogan committed:
docs(vec_approx_equal): clean up some outdated/misleading comments

brookslogan committed:
Update is_locf replacement test for new NA vs. NaN results

brookslogan committed:
docs(vec_approx_equal): update wrt compatibility fixes

brookslogan committed:
refactor(epi_slide_opt.epi_archive): address TODOs

- Column name clobbering is already checked for / avoided in main function code.
- Move input nesting and output rbinding inside of
  epi_slide_opt_archive_one_epikey.

brookslogan committed:
Update R/epi_slide_opt_edf.R

Co-authored-by: nmdefries <42820733+nmdefries@users.noreply.github.com>

brookslogan committed:
Update R/epi_slide_opt_edf.R

Co-authored-by: nmdefries <42820733+nmdefries@users.noreply.github.com>

brookslogan committed:
Update R/epi_slide_opt_edf.R

Co-authored-by: nmdefries <42820733+nmdefries@users.noreply.github.com>

brookslogan committed:
Update R/epi_slide_opt_edf.R

Co-authored-by: nmdefries <42820733+nmdefries@users.noreply.github.com>

brookslogan committed:
docs: document (GHA)

brookslogan committed:
docs: document (GHA)

brookslogan committed:
refactor(epi_slide_opt_archive): add classes to errors

brookslogan committed:
Skip malfunctioning test being updated separately

brookslogan commented on R/patch.R:
```diff
@@ -0,0 +1,348 @@
+  # Most input validation + handle NULL earlier_snapshot. This is a small function so
+  # use faster validation variants:
+  if (!is_tibble(later_tbl)) {
+    cli_abort("`later_tbl` must be a tibble")
```
Added!

brookslogan commented on R/epi_slide_opt_archive.R:
```diff
@@ -0,0 +1,221 @@
+        group_updates <- group_values %>%
+          nest(.by = version, .key = "subtbl") %>%
+          arrange(version)
+        # TODO move nesting inside the helper?
```
Should be addressed now.

brookslogan committed:
style: styler (GHA)

nmdefries committed:
Use data.table column looping in epi_slide_opt.epi_archive

brookslogan committed:
Avoid `map` overhead in loop

brookslogan committed:
Work toward proper `fill =` + `.align` support in `epi_slide_opt()`

This approach to getting the `fill` value makes things make more sense
internally, but we still may need to manually deal with the impact of our own
padding and directly intercept or disallow `fill =`.

Suggested-by: nmdefries <nmdefries@users.noreply.github.com>

brookslogan committed:
Don't try to support `epi_slide_opt(fill =)`

brookslogan committed:
docs: document (GHA)

brookslogan committed:
docs(epi_slide_opt): update comments re. after>=1 padding relocation

brookslogan commented on R/epi_slide_opt_archive.R:
```diff
@@ -0,0 +1,221 @@
+      }
+    } else {
+      cli_abort("epiprocess internal error: `f_from_package` was {format_chr_deparse(f_from_package)}")
+    }
```
Incorporated!  (Diff was outdated so I needed to manually fiddle with it.)

brookslogan committed:
Address styler & linter indentation conflict

dsweber2 committed:
docs: document (GHA)

dsweber2 reviewed (commented):
Not done with my review by any means, but I think it probably makes more sense to do a comment at a time instead of bundling them all.

dsweber2 commented on R/epi_slide_opt_archive.R:
```diff
@@ -0,0 +1,230 @@
+    # adaptive at tails + perform any NA gap-filling needed:
+    slide <- inp_snapshot[slide_inp_backrefs, ]
+    slide$time_value <- slide_time_values
+    if (f_from_package == "data.table") {
```
This is the only branch that's actually tested right?

brookslogan replied:
Good catch!  I've done some ad-hoc testing with other functions but actual tests are looking pretty homogeneous...  Should probably look into code coverage tools again.

dsweber2 commented on R/epi_slide_opt_archive.R:
```diff
@@ -0,0 +1,230 @@
+      }
+    out_update <- slide[rows_should_keep, ]
+    out_diff <- tbl_diff2(prev_out_snapshot, out_update, "time_value", "update")
+    out_snapshot <- tbl_patch(prev_out_snapshot, out_diff, "time_value")
+    prev_inp_snapshot <<- inp_snapshot
+    prev_out_snapshot <<- out_snapshot
```
```suggestion
    prev_inp_snapshot <<- inp_snapshot
    prev_out_snapshot <<-  tbl_patch(prev_out_snapshot, out_diff, "time_value")
```

dsweber2 replied:
`out_snapshot` isn't used anywhere else?

dsweber2 commented on R/patch.R:
```diff
@@ -0,0 +1,375 @@
+#' @return a tibble in compact "update" (diff) format
+#'
+#' @keywords internal
+tbl_diff2 <- function(earlier_snapshot, later_tbl,
```
not really important but why isn't this just `tbl_diff`?

brookslogan replied:
To attempt to clarify that it's a binary operation, unlike `diff`.

dsweber2 replied:
right this is a thing that e.g. purrr uses

dsweber2 commented on R/patch.R:
```diff
@@ -0,0 +1,375 @@
+
+  if (later_format == "update") {
+    # Cases 4. and 5.:
+    combined_tbl <- combined_tbl[combined_from_later & !combined_compactify_away, ]
```
Case 1 isn't relevant here because it's indistinguishable from unchanged rows when we're just updating, right?

dsweber2 commented on R/epi_slide_opt_archive.R:
```diff
@@ -0,0 +1,230 @@
+epi_slide_opt.grouped_epi_archive <- function(.x, ...) {
+  assert_set_equal(
+    group_vars(.x),
+    key_colnames(.x$private$ungrouped, exclude = c("time_value", "version"))
```
I guess we ban grouping by non-keys? That does make sense, first time I think I've actually run into it though.

dsweber2 replied:
actually would this ban grouping by a subset of key_colnames? that seems like something that we would want to support

brookslogan replied:
Like with the current `epi_slide_opt.epi_df`, we require the grouping to be by the epikeys.  There are some frustrating situations to think about when grouping by other things, such as scattered location drop-in & drop-out & how we should do completion.

dsweber2 commented on R/epi_slide_opt_archive.R:
```diff
@@ -0,0 +1,230 @@
+        if (use_progress) cli::cli_progress_update(id = progress_bar_id)
+        res
+      }) %>%
+      as.data.frame() %>% # data.table#6859
```
This issue? looks like they fixed it. https://github.com/Rdatatable/data.table/issues/6859

I guess you're trying to avoid having to pin data.table to a recent version?

brookslogan replied:
I filed this issue as I was writing this PR.  We could do a version range requirement I guess; has this hit CRAN?

dshemetov replied:
It's not in the 1.17.0 release, will need to wait a bit.

dsweber2 commented on R/epi_slide_opt_edf.R:
```diff
@@ -0,0 +1,705 @@
+      time_unit_abbr <- time_type_unit_abbr(time_type)
+      align_abbr <- c(right = "", center = "c", left = "l")[[.align]]
+    }
+    glue_env <- rlang::env(
```
This whole section seems overly indirect, but I guess it's so that `glue` can allow the user to set the prefix to e.g. `"{n}^{time_unit_abbr}"`  themselves or something like that?

brookslogan replied:
Something like that (e.g., `"_{.n}{.time_unit_abbr}_mean"`).  Idea was that if you're copy-pasting & adjusting epi_slide_opts, or tweaking epi_slide_opt parameters over time, it might be more conveniently correct.  Also originally it was intended to prevent "magical defaults" but not sure that was actually a problem.

brookslogan replied:
(This naming code was part of an earlier PR; it appears here because I put it inside of a helper function & in a new file.)

dsweber2 commented on R/patch.R:
```diff
@@ -0,0 +1,375 @@
+    res <- fifelse(
+      !na_or_nan1 & !na_or_nan2,
+      abs(vec1 - vec2) <= abs_tol,
+      if (na_equal) {
```
doesn't this break the entire point of `fifelse` by forcing it to use an R for loop?

dsweber2 commented on R/patch.R:
```diff
@@ -0,0 +1,375 @@
+    na_or_nan1 <- is.na(vec1)
+    na_or_nan2 <- is.na(vec2)
+    # Since above are bare logical vectors, we can use `fifelse`
+    res <- fifelse(
+      !na_or_nan1 & !na_or_nan2,
+      abs(vec1 - vec2) <= abs_tol,
+      if (na_equal) {
+        na_or_nan1 & na_or_nan2 & (is.nan(vec1) == is.nan(vec2))
+      } else {
+        # Like `==` and `vec_equal`, we consider NaN == {NA, NaN, anything else}
+        # to be NA.
+        NA
+      }
+    )
```
```suggestion
    if (na_equal) {
      res <- fifelse(
        !na_or_nan1 & !na_or_nan2,
        abs(vec1 - vec2) <= abs_tol,
        na_or_nan1 & na_or_nan2 & (is.nan(vec1) == is.nan(vec2))
      )
    } else {
      # Like `==` and `vec_equal`, we consider NaN == {NA, NaN, anything else}
      # to be NA.
      res <- fifelse(
        !na_or_nan1 & !na_or_nan2,
        abs(vec1 - vec2) <= abs_tol,
        NA
      )
    }
```

dsweber2 replied:
doesn't this break the entire point of fifelse by forcing it to use an R for loop?

brookslogan replied:
Don't think there's an R loop here.

Is suggestion for clarity or intended for performance?  (For clarity, what would you think about 
```r
      if (na_equal) {
        na_comparisons <- na_or_nan1 & na_or_nan2 & (is.nan(vec1) == is.nan(vec2))
      } else {
        # Like `==` and `vec_equal`, we consider NaN == {NA, NaN, anything else}
        # to be NA.
        na_comparisons <- NA
      }
      fifelse(
        [...],
        na_comparisons
      )
```
?)

dsweber2 replied:
For performance primarily. `fifelse` is inherently mapping across a vector of values. We're interrupting that mapping using a base R `if`. The alternative you're suggesting also seems fine to me.

brookslogan replied:
I don't think there should be a performance difference.  I think the way this works out is
fifelse(test, yes, no)
- sets up a function evaluation environment with a, b, and c being delayed bindings / promises
- hits `.Call(CfifelseR, test, yes, no, na)`; FFI forces all arguments, causing the `no` promise to be evaluated (into a logical vector).
- FFI continues merrily on its way

But I don't see a problem with writing it this other way either.  Thanks for the edits in your PR!

dsweber2 replied:
So I did some direct testing on this because I'm a bit confused
```r
  n_entries <- 1e8
  na_or_nan1 <- sample(c(TRUE, FALSE), n_entries, TRUE)
  na_or_nan2 <- sample(c(TRUE, FALSE), n_entries, TRUE)
  vec1 <- runif(n_entries)
  vec1[na_or_nan1] <- NA
  vec2 <- runif(n_entries)
  vec2[na_or_nan2] <- NA
  na_equal <- TRUE
  abs_tol <- 1e-8
  system.time(
    res <- fifelse(
      !na_or_nan1 & !na_or_nan2,
      abs(vec1 - vec2) <= abs_tol,
      if (na_equal) {
        na_or_nan1 & na_or_nan2 & (is.nan(vec1) == is.nan(vec2))
      } else {
        NA
      }
    )
  )
  system.time(
    {if (na_equal) {
      na_comparisons <- na_or_nan1 & na_or_nan2 & (is.nan(vec1) == is.nan(vec2))
    } else {
      na_comparisons <- NA
    }
    res <- fifelse(
         !na_or_nan1 & !na_or_nan2,
         abs(vec1 - vec2) <= abs_tol,
         na_comparisons
    )}
  )
```
Gives pretty comparable times of ~2.9s. I'm mostly confused how the `no` case of 
```r
      if (na_equal) {
        na_or_nan1 & na_or_nan2 & (is.nan(vec1) == is.nan(vec2))
      } else {
        NA
      }
```
is actually vectorized here. I guess it's only applying the case once, rather than in parallel.

brookslogan replied:
Let N be the length of the `test` vector.  `fifelse` evaluates `yes` and `no` when feeding it into its native function, expecting each to be either length-N or length-1; in the common `length(yes) == N, length(no) == N` case, it just straight up expects to be provided N+N values of `yes` and `no`, half of which are going to be thrown away due to `test` selecting them away.  (Hypothetically, we could try avoiding calculating values of `yes` and `no` that we know will be thrown away & ditch `fifelse`, but it probably won't look as nice and I don't know whether or not to expect it to actually be faster.)

If `na_equal` is `TRUE`, then we are feeding in a length-N vector for `yes` and a length-N vector for `no`; all the operations used to build the `no` vector are already vectorized (`&`, `is.nan`, `==`).  If `na_equal` is `FALSE`, then we are feeding in a length-N vector for `yes` and a length-1 vector for `no`; `fifelse` recycles the `no` vector (at least conceptually) to match the length of `test`.

dsweber2 commented on R/patch.R:
```diff
@@ -0,0 +1,375 @@
+        vec_approx_equal0(vec1[[col_i]], vec2[[col_i]], na_equal, abs_tol, inds1, inds2)
+      }))
+    }
+  } else if (is_bare_list(vec1) && abs_tol != 0) {
```
I don't think this branch is getting tested at all.

dsweber2 commented on tests/testthat/test-archive.R:
```diff
@@ -92,7 +92,7 @@ test_that("epi_archives are correctly instantiated with a variety of data types"
 
   ea2 <- as_epi_archive(df, other_keys = "value", compactify = FALSE)
-  expect_equal(key(ea2$DT), c("geo_value", "time_value", "value", "version"))
+  expect_equal(key(ea2$DT), c("geo_value", "value", "time_value", "version"))
```
where did this order change come from? It does seem like an odd order; why isn't it geo_value, time_value, version, value?

brookslogan replied:
This probably comes from us moving from geo_value, time_value, other_keys[, version] canonical ordering to geo_value, other_keys, time_value[, version].  It looks pretty weird here probably moreso due to using the column name "value" rather than "age_group" or something like that.

brookslogan replied:
Sorry, actually we're not canonically ordering archive cols themselves in `as_epi_archive`, but are at least using the new canonically ordering of the vars in the key.  Change presumably comes from
```r
key_vars <- c("geo_value", other_keys, "time_value", "version")
```
line, but it's mixed up in some changes addressing data.table footguns so it's not quite obvious.

dsweber2 commented on tests/testthat/test-compactify.R:
```diff
@@ -124,7 +124,7 @@ quantile_pred_once <- function(estimates_vec, levels_vec) {
 }
 test_that("compactify works on distributions", {
-  skip("Until #611 is merged or hardhat/epipredict is patched")
+  skip("See #631.")
```
weird. Dan wrote both, so you may just want to talk to him about it

brookslogan replied:
Chatted with him a bit.  Submitted a PR to hardhat to fix this + some other issues & then we can remove a partial fix sitting in epipredict.

dsweber2 commented on tests/testthat/test-epi_slide_opt_archive.R:
```diff
@@ -0,0 +1,219 @@
+      epi_slide_opt(case_rate, frollmean, .window_size = 7)
+  )
+
+  mini_case_death_rate_archive_time_gen <- system.time(
```
So comparing these times, it's ~100x faster? Is this representative of other experiments you've done? That's certainly pretty nice if it is.

dsweber2 commented on tests/testthat/test-epi_slide_opt_archive.R:
```diff
@@ -0,0 +1,219 @@
+      group_by(geo_value)
+  )
+
+  mini_case_death_rate_archive_b <- mini_case_death_rate_archive %>%
+    {
+      as_tibble(as.data.frame(.$DT))
+    } %>%
```
```suggestion
  mini_case_death_rate_archive_b <-
    mini_case_death_rate_archive$DT %>%
    as.data.frame() %>%
    as_tibble()

```

dsweber2 commented on tests/testthat/test-epi_slide_opt_archive.R:
```diff
@@ -0,0 +1,219 @@
+      epi_slide_opt(percent_cli, frollmean, .window_size = 7)
+  )
+
+  archive_cases_dv_subset_time_gen <- system.time(
```
so this pair is a factor of ~11

dsweber2 committed:
comments and some nit rewrites

dsweber2 reviewed (changes requested):
So I've got some qualms about the complexity and the relatively narrow set of tests, but it does seem to be working, and would be useful for porting a faster `daily_to_weekly` transform.

There's a number of nits which I'd appreciate you fixing, and I've made a PR with some notes and some of the nits changed

dsweber2 committed:
docs: document (GHA)

brookslogan commented on R/archive.R:
```diff
@@ -464,56 +462,38 @@ update_is_locf <- function(arranged_updates_df, ukey_names, abs_tol) {
-      abs(vec - lag_vec) <= abs_tol,
-      is.na(vec) & is.na(lag_vec)
-    ))
+  n_updates <- nrow(arranged_updates_df)
```
Nice catch; this is outdated.  Before, we did use it within a `filter`, and we used `if_all`, which required dplyr-data-masking.

todo: remove

dsweber2 replied:
it's removed in the PR on this

brookslogan commented on R/archive.R:
```diff
@@ -536,11 +516,32 @@ as_epi_archive <- function(
   x <- rename(x, ...)
-  x <- guess_column_name(x, "time_value", time_column_names())
   x <- guess_column_name(x, "geo_value", geo_column_names())
+  if (!all(other_keys %in% names(x))) {
```
No behavior-driven reason.  We previously canonically ordered columns `geo_value, time_value, other_keys, [version,] <values>`; but have moved toward `geo_value, other_keys, time_value, [version,] <values>` in various places; I was just trying to match that.

dsweber2 replied:
Consistency is more important than the specific order. Good to know

brookslogan commented on R/epi_slide_opt_archive.R:
```diff
@@ -0,0 +1,230 @@
+  unit_step <- unit_time_delta(time_type)
+  prev_inp_snapshot <- NULL
+  prev_out_snapshot <- NULL
+  result <- map(seq_len(nrow(grp_updates_by_version)), function(version_i) {
```
Right, that's the idea.  We also get the time range that was updated, allowing us to limit slide computation time ranges.  Comparing progressive patching vs. `epix_as_of`... I thought I did at some point, but now I'm questioning whether I did it in the right way, plus I've sped up some other internals.  Trying to do a quick comparison for 7dav-ing now:

- Splitting into tibbles for each epikey vs. operating on combined epi_dfs has 3.8x the speed in `archive_cases_dv_subset`, 0.8x on `case_death_rate_archive`.

The rest of this comparison is based on combined epi_dfs, so it may not be completely comparable.
- The special time logic actually slows things down on these two test cases, ... 0.82x / 0.93x speed of without.  However, this definitely didn't seem to be the case during development, so either some "incidental" difference or other improved internal made a difference.  These data sets are also probably not the best setup for the special time logic, as they actually have pretty large lag ranges recorded in updates.
- The progressive-patching to get inputs is 1.32x / 1.12x faster than `epix_as_of` querying, given that the time range logic has been removed.
- The progressive-diffing of the output is comparable to rbinding + compactifying all at once in terms of speed, but should avoid some potential RAM issues when dealing with larger archives; given all features above removed.
- Using dedicated code here vs. delegating to `epix_slide` is 1.14x / 31x (!) faster than `epix_slide`.

So... potential responses here: I could try
- A. removing the special time range logic and see the impact on current code.  (Potentially simpler & faster!  Or could point to something yet more complicated but yet faster.)
- B. removing the patching logic (might slow things down non-negligibly based on the above).
- C. locating the `epix_slide` overhead.
- D. rewriting `epix_slide` with progressive-patching logic.
- E. rewriting this code and eventually `epix_slide` with iterator code that separates the progressive-patching, sliding, and progressive-diffing code into more readable, re-usable chunks.
- F. try again to simplify `tbl_diff2`... this is likely very hard to do with `compactify_abs_tol` involved, though I guess I never actually used that feature in `epi_slide_opt`.  If we keep that feature out, then I think I probably have something faster & simpler sitting around.

Right now, I'm thinking maybe I should take a look at A here, and we should decide whether `compactify_abs_tol` should be supported here and whether you're willing to review a reimplementation from F.  The others... I'm not sure B. is that much of a simplification if progressive diffing is kept.  I'd like to do C. and/or E.+D. at some point, but separately.

Thoughts?

dsweber2 replied:
> Trying to do a quick comparison for 7dav-ing now:

Could you share what you're comparing to in some way? If it is the case that we're only getting a 1.3x speed-up relative to e.g. a for loop+`epix_as_of`, I don't think the level of complexity we're adding in this PR is worth it. The 10-100x that show up in the tests comparing vs `epix_slide(epi_slide_opt())` definitely seems worthwhile, but if that's coming from potential improvements to `epix_slide`, it'd be better to re-write this to use `epix_slide` and improve that in a different PR.

brookslogan replied:
The 1.3x is comparing [this](https://github.com/brookslogan/scratchpads/blob/bf0c1dae3ca8ef8368f26f8fda13e10b61799e9c/archive-agg/archive-agg.R#L2816) to [this](https://github.com/brookslogan/scratchpads/blob/bf0c1dae3ca8ef8368f26f8fda13e10b61799e9c/archive-agg/archive-agg.R#L2827).  

Here, `map_ea` is akin to `map_dfr`, but instead of collecting & rbinding the results, it
- Immediately diffs each result with the previous one.  The source of the desired max RAM usage improvement.
- Collects the diffs.
- Rbinds the diffs and converts to an archive.
However, this version of `map_ea` is not backed by `tbl_diff2` but an earlier `epi_diff2`, which is slower.

dsweber2 replied:
ok, I misunderstood, I thought the 1.3x you were talking about was for the entire set of optimizations, and not just `epix_as_of` vs `tbl_patch`. I'm not sure whether leaving it as is, or trying to pare back to a minimal set of improvements is preferable. On the one hand, maintaining the 2.3k loc in this PR is going to be a lot. On the other it would require yet more re-writes to pare this back. to a minimal set

brookslogan replied:
Trying to break down LOC/doc (GitHub doesn't make this easy...) + collect some thoughts about where redundant implementation can be trimmed while keeping the same functions available + a few incomplete thoughts about paring functions down / obstacles.
```
<=230 epi_slide_opt.epi_archive
>=220 test-epi_slide_opt_archive.R
<=184 vec_approx_equal
<= 96 epi_slide_opt.epi_df (epi_slide_opt_edf.R + slide.R - old slide.R = 707 + 550 - 1161)
   79 tbl_diff2
   59 tbl_patch
>= 57 test-vec_approx_equal
   52 tbl_fast_anti_join

(can maybe get -66-34 on epi_slide_opt.epi_df or -138 epi_archive)
(can maybe get -31 on vec_approx_equal by removing inds{1,2})
(need to add tests)
(can maybe get -184-57 by contributing vec_approx_equal upstream... though likely adding a bunch of work)
(can maybe get -52 by contributing tbl_fast_anti_join upstream... though likely adding a bunch of work) 
(can maybe get very roughly -57 by making apply_compactify use tbl_diff2, if perf good enough)

(minimal slower & RAM-exploding epi_slide_opt.epi_archive likely maybe ~27 lines in epi_slide_opt.epi_archive)
(load-archives-from-snapshot-files will require tbl_diff2, tbl_fast_anti_join, vec_approx_equal... or a backend redesign)
```

- [ ] todo: @return for epi_slide_opt.epi_archive

brookslogan committed:
Update R/epi_slide_opt_archive.R

Co-authored-by: David Weber <david.weber2@pm.me>

brookslogan committed:
style: styler (GHA)

brookslogan committed:
Update tests/testthat/test-epi_slide_opt_archive.R

Co-authored-by: David Weber <david.weber2@pm.me>

brookslogan committed:
Update R/epi_slide_opt_archive.R

brookslogan committed:
Update R/patch.R

brookslogan committed:
Update R/patch.R

brookslogan committed:
docs: document (GHA)

brookslogan committed:
Merge pull request #652 from cmu-delphi/archive-agg-nits

comments and some nit rewrites

brookslogan committed:
Simplify `tbl_diff2` with `tbl_fast_anti_join`

brookslogan committed:
fix(tbl_fast_anti_join): include non-ukey, non-val cols in result

brookslogan committed:
Comment, clean, style, fix @keywords in archive opt slide & helpers

brookslogan committed:
docs: document (GHA)

brookslogan committed:
Fix incomplete rename refactor

brookslogan committed:
perf(epi_slide_opt.epi_archive): more `[` -> `vec_slice` changes

14% or 17% time reduction for 7daving on a couple of test archives.

brookslogan committed:
docs(patch.R): fix comment grammar

brookslogan commented on R/patch.R:
Maybe.  Probably would need some generalization work then.  The `vec_` prefix is, as in `vctrs`, to denote that the first arg is a vctrs-sense vector, and to distance further from `all.equal`, which returns a scalar result.

brookslogan committed:
fix: add missing importFrom

brookslogan commented on R/epi_slide_opt_archive.R:
```diff
@@ -0,0 +1,221 @@
+      checkmate::check_logical(.progress, any.missing = FALSE, len = 1L, names = "unnamed"),
+      checkmate::check_string(.progress)
+    )
+    if (isTRUE(.progress)) {
+      .progress <- "Time series processed:"
+    }
+    use_progress <- !isFALSE(.progress)
```
This would not use a progress bar if the user passed their own progress bar label string.

- [ ] todo: docs.  Maybe tests if that is possible?

brookslogan committed:
Refactor and speed up branch of `vec_approx_equal` NA/NaN testing

brookslogan committed:
Test unifying epi_slide_opt inner comps between edf and archive

brookslogan committed:
Refactor unified epi_slide_opt_one_epikey for clarity

brookslogan committed:
Clean up slide range logic

brookslogan committed:
perf(unit_time_delta): faster arg matching

brookslogan committed:
refactor: handle time +/- Inf in helper function

brookslogan committed:
Attempt rewriting combined opt slide logic to match old archive perf chars

Something still seems different; some Date arithmetic stuck out, and attempts to
avoid helped somewhat, but this still seems slower for archives than pre-unified
approach.

brookslogan committed:
perf: try `vec_c` -> `c` on 3 time_values

As `vec_c` is slower on 3 scalar `Date`s, and it's probably pretty uncommon to
use `time_value`s with `c` incompatible with `vec_c` (though possible... maybe
tibbles with year, week, wday, like from MMWRweek; though these would probably
break inside archive DT).

brookslogan committed:
perf: avoid `c.Date` when possible

Revert earlier change to try to reduce time inside `time_minus_time_in_n_steps`,
as `c.Date` is more costly.  This also works back towards generality from the
previous `vec_c` -> `c` change.

brookslogan committed:
perf: old, faster slide edge-trimming args are actually usable

brookslogan committed:
fix(time_plus_slide_window_arg): on non-integerish, non-Inf `y`

brookslogan committed:
fix: output filter w/ before=Inf, wrong before/after types

Also change a map -> lapply so we can get srcref to internal error immediately,
add a missed .window_size-missingness check, delete some helper functions that
were sort of helpful but also tacked on significant run time when used in a
natural way.

brookslogan committed:
refactor(epi_slide_opt): helper function&variable naming, arg ordering

brookslogan committed:
More internal renames and documentation

brookslogan committed:
perf: c -> list when specifying slide out date min, max

brookslogan committed:
fix(epi_slide_opt): on `out_filter_time_set` narrowing to 0

brookslogan committed:
docs(slide.R): correct error message regarding data.frame rownames

The error message says that slide computations can't be rownamed data frames.
This is both bad wording (having rownames != having non-automatic rownames) and
we're not actually enforcing any restriction of the sort anyway. We drop/ignore
the rownames though (via dplyr/tibble). In `epix_slide`, the allowance of
non-automatic rownames was deliberate; in `epi_slide`, it also seems likely the
way to go; one might convert to `data.frame` and have some filter-like operation
return non-automatic rownames.

brookslogan committed:
fix: `purrr::partial` missing future arg placement

Use `... =` to specify this.  Prevents some confusing error messages on
```
archive_cases_dv_subset %>%
  epi_slide_opt(percent_cli, frollmean, 0, .window_size = 7)
```
but does not solve all issues that can happen with unnamed `...`.
